sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExec.scala

```diff
@@ -181,6 +217,142 @@ class GpuWindowExecMeta(windowExec: WindowExec,
+        val attr = postRemapping(a.exprId).head._2
+        GpuAlias(child, attr.name)(attr.exprId, attr.qualifier)
+      case other => other
+    }
+  }
+
+  private def hasGpuWindowFunction(expr: Expression): Boolean = {
+    expr.find {
+      case _: GpuWindowExpression => true
+      case _ => false
+    }.isDefined
+  }
+
+  private def extractAndSave(expr: Expression,
+      saved: ArrayBuffer[NamedExpression],
+      deduped: mutable.HashMap[Expression, Attribute]): Expression =
+    expr match {
+      // Don't rename an already named expression
+      case ne: NamedExpression =>
+        val missingExprId = Set(ne.exprId) -- saved.map(_.exprId)
+        if (missingExprId.nonEmpty) {
+          saved += ne
+        }
+        ne.toAttribute
+      case e: Expression if e.foldable =>
+        e // No need to create an attribute reference if it will be evaluated as a Literal.
+      case e: Expression =>
+        // For other expressions, we extract it and replace it with an AttributeReference (with
+        // an internal column name, e.g. "_gpu_w0"). Deduping it as we go.
+        deduped.getOrElseUpdate(e, {
+          val withName = GpuAlias(e, s"_gpu_w${saved.length}")()
+          saved += withName
+          withName.toAttribute
+        })
+    }
+
+  /**
+   * In some distributions expressions passed into WindowExec can have more operations
+   * in them than just a WindowExpression wrapped in an GpuAlias. This is a problem if we
+   * want to try and do multiple window operations in a single pass to speed things up
+   * or if we need to add new transitive window functions when we are doing some memory
+   * optimizations, like running window. This will split the input expressions
+   * into three sets of expressions. The first set is a project with no window expressions in it at
+   * all. The second takes the first as input and will only have aliases to columns in the first or
+   * named expressions wrapping a single window function in it. The third uses the second as
+   * input and will do any final steps to combine window functions together.
+   *
+   * For example `SUM(a) - SUM(b + c) over (PARTITION BY d ORDER BY e) as result` would be
+   * transformed into
+   * <pre>
+   * Phase 1 (Pre project):
+   * a, b + c as _tmp0, d, e
+   *
+   * Phase 2 (Window Operations):
+   * SUM(a) over (PARTITION BY d ORDER BY e) as _tmp1,
+   * SUM(_tmp0) over (PARTITION BY d ORDER BY e) as _tmp2
+   *
+   * Phase 3 (Post Project):
+   * (_tmp1 - _tmp2) as result
+   * </pre>
+   *
+   * This assumes that there is not a window function of another window function, like
+   * `LAG(SUM(a), 2)` which appears to be something all distros split apart into separate
+   * window operations, so we are good.
+   * @param exprs the input expressions to a GpuWindowExec
+   */
+  def splitAndDedup(exprs: Seq[NamedExpression]):
+  (Seq[NamedExpression], Seq[NamedExpression], Seq[NamedExpression]) = {
```

Comment:
Nit: Instead of a raw tuple consider using a private `case class` to track output parameters.

Reply:
I just don't see it for a single API that is going to wrap a return result and then unwrap it immediately.

```
val (pre, windowOps, post) = GpuWindowExec.splitAndDedup(gpuWindowExpressions)
```

vs

```
GpuSplitWindowExpressions(pre, windowOps, post) = GpuWindowExec.splitAndDedup(gpuWindowExpressions)
```

Scala will already hide the names when creating it, and likely unapplying it like I would probably do. If it was going to be passed around a bit more, then I would understand, but I don't see it here.